lib/src/util/tested_expressions.dart

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.util.tested_expressions;
+
+import 'dart:collection';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:linter/src/util/boolean_expression_utilities.dart';
+
+typedef void _recurseCallback(Expression expression);
+
+class TestedExpressions {
+  final BinaryExpression testingExpression;
+  final LinkedHashSet<Expression> truths;
+  final LinkedHashSet<Expression> negations;
+  LinkedHashSet<ContradictoryComparisons> _contradictions;
+
+  TestedExpressions(this.testingExpression, this.truths, this.negations);
+
+  LinkedHashSet<ContradictoryComparisons> evaluateInvariant() {
+    if (_contradictions != null) {
+      return _contradictions;
+    }
+
+    _contradictions = _findContradictoryComparisons(
+        new LinkedHashSet.from(
+            [testingExpression.leftOperand, testingExpression.rightOperand]),
+        testingExpression.operator.type);
+
+    if (_contradictions.isEmpty) {
+      HashSet<Expression> set = _extractComparisons(testingExpression)
+        ..addAll(truths)..addAll(negations);
+      // Here and in several places we proceed only for
+      // TokenType.AMPERSAND_AMPERSAND because we then know that all comparisons
+      // must be true.
+      _contradictions.addAll(
+          _findContradictoryComparisons(set, TokenType.AMPERSAND_AMPERSAND));
+    }
+
+    return _contradictions;
+  }
+
+  /// TODO: A truly smart implementation would detect
+  /// (someComputation() && otherComputation()) &&
+  /// (!someComputation() || !otherComputation())
```

Comment:
What if the computations have side-effects that cause them to return different values from one call to the next?